keras/src/saving/keras_saveable.py

```diff
@@ -30,9 +37,14 @@ def __reduce__(self):
         keras saving library."""
         import keras.src.saving.saving_lib as saving_lib
 
-        buf = io.BytesIO()
-        saving_lib._save_model_to_fileobj(self, buf, "h5")
+        model_buf = io.BytesIO()
+        saving_lib._save_model_to_fileobj(self, model_buf, "h5")
+
+        custom_objects_buf = io.BytesIO()
+        pickle.dump(get_custom_objects(), custom_objects_buf)
+        custom_objects_buf.seek(0)
+
         return (
             self._unpickle_model,
-            (buf,),
+            (model_buf, custom_objects_buf),
```

Comment:
This would break existing pickles, no?

Reply:
Good catch, I have not seen this because I use `cloudpickle` instead of `pickle`.

I confirm that it breaks compatibility if users used the standard `pickle` lib to dump their previous models. However, previous models dumped with `cloudpickle` do not have such compatibility issue because the definition of the unpickling callable is directly stored inside the file (not just a reference).

Handling compatibility issues will probably not be straightforward. Not sure it is worth the effort.

Reply:
The compatibility issue could be handled by doing something like:

```python
    @classmethod
    def _unpickle_model(cls, model_buf, *args):
        import keras.src.saving.saving_lib as saving_lib

        # pickle is not safe regardless of what you do.

        if len(args) == 0:
            return saving_lib._load_model_from_fileobj(
                model_buf,
                custom_objects=None,
                compile=True,
                safe_mode=False,
            )

        else:
            custom_objects_buf = args[0]
            custom_objects = pickle.load(custom_objects_buf)
            return saving_lib._load_model_from_fileobj(
                model_buf,
                custom_objects=custom_objects,
                compile=True,
                safe_mode=False,
            )
```

But after experimenting with `pickle` instead of `cloudpickle`, I do not think this PR is adding value if we stick with the standard `pickle` because pickling the custom object only returns a reference in that case (meaning we still need the object to be available at inference time). 

2 options:
* We pickle the custom objects only if `cloudpickle` is importable (without explicitly adding it as a dependency)
* We close the PR now

Reply:
New version with the following changes:
* Use `cloudpickle` in the `__reduce__()` method but keep `pickle` in the unpickling class method so that we only need `cloudpickle` in the training environment (possible because `cloudpickle.load` is just an alias of `pickle.load`)
* Use a dict instead of a tuple in the arguments for backward compat support
* Better isolate tests by adding a pytest fixture to clean the custom objects global dict before each test runs (failing tests in previous commit were due to already registered (but not accessible) custom objects by other tests
* Add a test running in a fake __main__ module (so that cloudpickle could serializes the objects) and delete the custom object variable before loading the pickle to verify that its definition is completely stored inside the file